modules/apps/oauth2-provider/oauth2-provider-web/src/main/resources/META-INF/resources/js/OAuth2Client.ts

```diff
@@ -0,0 +1,249 @@
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+// @ts-nocheck
+
+import pkceChallenge from 'pkce-challenge';
+
+class OAuth2Client {
+	authorizeURL: any;
+	clientId: any;
+	encodedRedirectURL: string;
+	homePageURL: any;
+	redirectURIs: any[];
+	tokenURL: any;
+
+	static FromParameters({
+		clientId,
+		homePageURL,
+
+		// @ts-ignore
+
+		authorizeURL = Liferay.OAuth2.getAuthorizeURL(),
+
+		// @ts-ignore
+
+		redirectURIs = [Liferay.OAuth2.getBuiltInRedirectURL()],
+
+		// @ts-ignore
+
+		tokenURL = Liferay.OAuth2.getTokenURL(),
+	}) {
+		const webClient = new OAuth2Client();
+
+		webClient.authorizeURL = authorizeURL;
+		webClient.clientId = clientId;
+		webClient.encodedRedirectURL = encodeURIComponent(redirectURIs[0]);
+		webClient.homePageURL = homePageURL;
+		webClient.redirectURIs = redirectURIs;
+		webClient.tokenURL = tokenURL;
+
+		return webClient;
+	}
+
+	static FromUserAgentApplication(userAgentApplicationName: string) {
+
+		// @ts-ignore
+
+		const userAgentApplication = Liferay.OAuth2.getUserAgentApplication(
+			userAgentApplicationName
+		);
+
+		if (!userAgentApplication) {
+			throw new Error(
+				`No Application User Agent profile found for ${userAgentApplicationName}`
+			);
+		}
+
+		const webClient = new OAuth2Client();
+
+		// @ts-ignore
+
+		webClient.authorizeURL = Liferay.OAuth2.getAuthorizeURL();
+		webClient.clientId = userAgentApplication.clientId;
+		webClient.encodedRedirectURL = encodeURIComponent(
+			userAgentApplication.redirectURIs[0]
+		);
+		webClient.homePageURL = userAgentApplication.homePageURL;
+		webClient.redirectURIs = userAgentApplication.redirectURIs;
+
+		// @ts-ignore
+
+		webClient.tokenURL = Liferay.OAuth2.getTokenURL();
+
+		return webClient;
+	}
+
+	async fetch(url: RequestInfo, options: any = {}): Promise<any> {
+		const oauth2Client = this;
+
+		return oauth2Client._fetch(url, options).then((response) => {
+			if (response.ok) {
+				const contentType = response.headers.get('content-type');
+				if (
+					contentType &&
+					contentType.indexOf('application/json') !== -1
+				) {
+					return response.json();
+				}
+				else {
+					return Promise.resolve(response);
+				}
+			}
+
+			return Promise.reject(response);
+		});
+	}
+
+	_createIframe(): HTMLIFrameElement {
+		const ifrm = document.createElement('iframe');
+		ifrm.style.display = 'none';
+		document.body.appendChild(ifrm);
+
+		return ifrm;
+	}
+
+	async _fetch(resource: RequestInfo, options: any = {}): Promise<any> {
+		const oauth2Client = this;
+
+		let resourceUrl: string = resource['url'] ? resource['url'] : resource;
```

Comment:
nit: You can just do `resource['url'] || resource`

Reply:
that gives me a typescript error.